File: Leetcode_Python/minRemoveToMakeValid.py
class Solution:
    def minRemoveToMakeValid(self, s: str) -> str:
        count = 0
        initialResult = []

        for c in s:
            if c == "(":
                initialResult.append(c)
                count += 1
            elif c == ")" and count > 0:
                initialResult.append(c)
                count -= 1
            elif c != ")":
                initialResult.append(c)

        finalResult = []

        for c in initialResult[::-1]:
            if c == "(" and count > 0:
                count -= 1
            else:
                finalResult.append(c)

        return "".join(finalResult[::-1])

File: Leetcode_Java/minRemoveToMakeValid.java
package Leetcode_Java;

class Solution {
    public String minRemoveToMakeValid(String s) {

        int count = 0;
        StringBuilder initialResult = new StringBuilder();

        for (char c : s.toCharArray()) {
            if (c == '(') {
                initialResult.append(c);
                count++;
            } else if (c == ')' && count > 0) {
                initialResult.append(c);
                count--;
            } else if (c != ')') {
                initialResult.append(c);
            }
        }

        StringBuilder finalResult = new StringBuilder();
        int openCount = count; 

        for (int i = initialResult.length() - 1; i >= 0; i--) {
            char c = initialResult.charAt(i);
            if (c == '(' && openCount > 0) {
                openCount--;
            } else {
                finalResult.append(c);
            }
        }

        return finalResult.reverse().toString();
    }
}